core/src/avm2/object/sound_object.rs

```diff
@@ -129,6 +137,89 @@ impl<'gc> SoundObject<'gc> {
         }
         Ok(())
     }
+
+    pub fn id3(self) -> Option<Object<'gc>> {
+        let this = self.0.read();
+        this.id3
+    }
+
+    pub fn set_id3(self, mc: &Mutation<'gc>, id3: Option<Object<'gc>>) {
+        let mut this = self.0.write(mc);
+        this.id3 = id3;
+    }
+
+    pub fn read_and_call_id3_event(self, activation: &mut Activation<'_, 'gc>, bytes: &[u8]) {
+        let id3 = activation
+            .avm2()
+            .classes()
+            .id3info
+            .construct(activation, &[])
+            .expect("failed to construct ID3Info object");
+        let tag = Tag::read_from(bytes);
+        if let Ok(ref tag) = tag {
+            if let Some(v) = tag.album() {
+                id3.set_public_property(
+                    "album",
+                    AvmString::new_utf8(activation.gc(), v).into(),
+                    activation,
+                )
+                .expect("failed set_public_property");
+            }
+            if let Some(v) = tag.artist() {
+                id3.set_public_property(
+                    "artist",
+                    AvmString::new_utf8(activation.gc(), v).into(),
+                    activation,
+                )
+                .expect("failed set_public_property");
+            }
+            if let Some(v) = tag.comments().next() {
+                id3.set_public_property(
+                    "comment",
+                    AvmString::new_utf8(activation.gc(), v.text.clone()).into(),
+                    activation,
+                )
+                .expect("failed set_public_property");
+            }
+            if let Some(v) = tag.genre() {
+                id3.set_public_property(
+                    "genre",
+                    AvmString::new_utf8(activation.gc(), v).into(),
+                    activation,
+                )
+                .expect("failed set_public_property");
+            }
+            if let Some(v) = tag.title() {
+                id3.set_public_property(
+                    "songName",
+                    AvmString::new_utf8(activation.gc(), v).into(),
+                    activation,
+                )
+                .expect("failed set_public_property");
+            }
+            if let Some(v) = tag.track() {
+                id3.set_public_property(
+                    "track",
+                    AvmString::new_utf8(activation.gc(), v.to_string()).into(),
+                    activation,
+                )
+                .expect("failed set_public_property");
+            }
+            if let Some(v) = tag.year() {
+                id3.set_public_property(
+                    "year",
+                    AvmString::new_utf8(activation.gc(), v.to_string()).into(),
+                    activation,
+                )
+                .expect("failed set_public_property");
+            }
+        }
+        self.set_id3(activation.context.gc_context, Some(id3));
```

Comment:
With this implementation, `id3` will always be non-null in FP- have you confirmed that this is true? Especially for mp3s without ID3 info

Reply:
Yep, it is always non-null, for audio that doesn't contain id3 info, the fields inside it will be null.